### ISSUE TITLE
UnManagedIndividuals Resources cannot be deserialized

### ISSUE THREAD

**Issue by cstockloew** [universAAL/middleware#39]
_Originally Opened: @Alfiva (2011-08-26 14:15:47_)
_Originally Closed: 2012-02-08 11:50:21_

I am using the mw.data.serialization features for the new version of the CHE.
I have a ContextEvent described in turtle RDF. I want to deserialize with the TurtleParser, but it cannot be: it is deserialized as a Resource.

The problem is in TurleParser.specialize(...), line 1143: type=types[0].
This means that, since ContextEvent is not a ManagedIndividual, it cannot be specialized. So, from all its types [Resource, Statement, ContextEvent], it is converted to the first one: Resource. I want it to be a ContextEvent! But also other future similar situations could be solved if instead of just selecting the first type, it selected the most specialized type (ContextEvent in my case).

To do so, I propose to include a getMostSpecializedClass method into Resource, just like the one there is on ManagedIndividual. In fact the code would be the same only that looking into uriRsrcClass instead of uriClassMap.

This way TurtleParser can assign the right type with type=r.getMostSpecialized... instead of directly assigning types[0].

--

From: _this issue has been automatically imported from our old issue tracker_

**Comment by cstockloew** [comment]
_Original author: @Alfiva (2011-09-12 12:55:45)_

(Note that I code references and lines are based on pre-osgi-extraction version. These may differ in current version)

**Comment by cstockloew** [comment]
_Original author: @cstockloew (2011-09-13 14:56:20)_

I cannot reproduce the problem. With the current version (0.3.2) this doesn't happen: ContextEvent only has one type (ContextEvent.MY_URI) and is blocking further types.
Please update and try again.

**Comment by cstockloew** [comment]
_Original author: @Alfiva (2011-11-03 11:40:35)_

Just as an update, I have tried with MW 0.3.1 as well and of course the problem still remains. I have also checked the TurtleParser code of 0.3.2 and didn´t find any measure that corrects this, so I assume that &quot;ContextEvent only has one type&quot; is due to how ContextEvent is constructed with the new 0.3.2 data repr. That ultimatley means that CHE is useless for retrieving rebuilt Context Events from the History, until it is adapted for 0.3.2...

**Comment by cstockloew** [comment]
_Original author: @cstockloew (2011-11-04 19:10:46)_

I downloaded the old version 0.3.0 (SVN rev 240) and it works. When I serialize a ContextEvent, the turtle representation looks as follows:

&lt;urn:org.universAAL.middleware.context.rdf:ContextEvent#_:c0a80001e1b6a342:1b&gt; a ns:ContextEvent ;

which is correctly only the ContextEvent. Also, the ContextEvent class is correctly registered (in static block: addResourceClass(MY_URI, ContextEvent.class);).

I'm using the following code snippet for testing (which I simply put in LightingServer.Activator):

	ServiceReference sr = context
		.getServiceReference(MessageContentSerializer.class.getName());
	if (sr == null)
	    return;

	MessageContentSerializer contentSerializer = (MessageContentSerializer) context
		.getService(sr);

	LightSource ls = new LightSource(&quot;testURI&quot;);
	ls.setLocation(new Room(&quot;testRoomURI&quot;));
	ls.setBrightness(0);
	org.universAAL.middleware.context.ContextEvent ce = new org.universAAL.middleware.context.ContextEvent(
		ls, LightSource.PROP_SOURCE_BRIGHTNESS);
	String str = contentSerializer.serialize(ce);
	LogUtils.logDebug(logger, &quot;Activator&quot;, &quot;start&quot;, new Object[] {
		&quot;\n      serialization dump (ContextEvent)\n&quot;, str, &quot;\n&quot; },
		null);

	org.universAAL.middleware.rdf.Resource r = (Resource) contentSerializer
		.deserialize(str);
	if (r != null)
	    LogUtils.logDebug(logger, &quot;Activator&quot;, &quot;start&quot;, new Object[] {
		    &quot;\n      serialization dump (ContextEvent)\n&quot;,
		    r.toStringRecursive(), &quot;\n&quot; }, null);

What I don't understand is the types that you have: [Resource, Statement, ContextEvent]. I cannot find the class &quot;Statement&quot; anywhere, it is undefined. In fact, ContextEvent is a direct subclass of Resource.

**Comment by cstockloew** [comment]
_Original author: @Alfiva (2011-11-07 09:28:40)_

The full strings of types in the &quot;types&quot; variable in the parser are in fact: [http://www.w3.org/2000/01/rdf-schema#Resource, http://www.w3.org/1999/02/22-rdf-syntax-ns#Statement, http://ontology.universAAL.org/Context.owl#ContextEvent].
They are RDF, not uAAL-specific, there is no need to be a class representing them.
The reason why you may not get these is because you are de/serializing only from/to uAAL. What CHE does is serializing from uAAL to Turtle, then deserializing from Turtle to Sesame store format. And then vice-versa when retrieving. Sesame uses its own Turtle parser, which explicits this chain of types. When retrieving the serialized form of an event from the store, it contains all its types, and this is what the uAAL Turtle parser does not recognize well.
In other words, I ´m relying on Turtle as the bridge &quot;language&quot; between uAAL and Sesame store, but the &quot;dictionaries&quot; uAAL-Turtle and Sesame-Turtle seem to differ in this matter. Which I expected to be similar, since AFAIK uAAL Turtle Parser is based on Sesame´s.
I assume then that this is not solved either in 0.3.2.

**Comment by cstockloew** [comment]
_Original author: @Alfiva (2012-01-04 13:34:50)_

Just a small follow up to keep the flame alive :)
Another explanation for this (althgouh I have not explicitly checked it) is that, OK, Context Events in uAAL have only the type Context Event. However when retrieving the events from the store, some degree of reasoning is enabled, so Sesame guesses that the context event is a:ContextEvent, which is a:Statement, which is a:Resource. The returned serialized event has the three types altogether.

I want to stress that this feature is important for CHE, otherwise its services that return Context Events are not working. The simplest workaround is just assigning the last value of types[] instead of the first one, types[0].

**Comment by cstockloew** [comment]
_Original author: @Alfiva (2012-02-03 10:11:39)_

Another follow up. I just want to stress that this fix is getting more and more necessary for CHe. Now I use also this functionality for the synchronization of mobile events too.
I have been using local modifications of mw.data.serialization, just changing types[0] to types[types.length-1] and it has worked OK in all scenarios so far, both for CHE and other normal operations.

**Comment by cstockloew** [comment]
_Original author: @cstockloew (2012-02-07 20:35:05)_

Should be done, please try again.

**Comment by cstockloew** [comment]
_Original author: @Alfiva (2012-02-08 11:50:21)_

OK, it works. Request closed.

Just as a closing word: Maybe it could be a good to put in TurtleParser

if(types.length==1)type=types[0]

This may help the performance, since most of the time this will be situation.

**Comment by cstockloew** [comment]
issue cloed on 2012-02-08 11:50:21
